### PR TITLE
Add Linting for MD files

### PR DESCRIPTION
**Pull request by akash-joshi**
This PR adds the markdown linter [`markdownlint`](https://www.npmjs.com/package/markdownlint-cli) so that we can check for correctness of any markdown file added/edited by any user.

Furthermore, I've also run its fixes one to fix all of the files we already have, and personally made changes to the `md` files where they couldn't be fixed by the linter automatically.

I also pushed these changes to `hasharchives` and tried building on that locally to check if there were any bugs introduced by these linting changes.

### CODE REVIEW

File: .github/workflows/markdown-linting/.markdownlint.json
Comment (akash-joshi) [inline .github/workflows/markdown-linting/.markdownlint.json:2]: These rules weren't compatible with how we're using the `md` files currently, or were blocking the PR from going live, so I've disabled them for now.

@kachkaev you can give these a look as well. We can fix the files manually based on the rules we enable in a follow-up task.

File: .github/workflows/mdx-lint.yml
Comment (akash-joshi) [inline .github/workflows/mdx-lint.yml:5]: Added to test it live.

File: resources/docs/simulation/creating-simulations/collaboration/README.md
Comment (nonparibus) [inline resources/docs/simulation/creating-simulations/collaboration/README.md:30]: @akash-joshi What rule results in this? Interesting

File: resources/docs/simulation/creating-simulations/configuration/topology/README.md
Comment (nonparibus) [inline resources/docs/simulation/creating-simulations/configuration/topology/README.md:113]:
```suggestion
* `euclidean_squared`: Similar to euclidean but distances are left squared for performance reasons
```

The other bullets don't have periods at the end.

File: resources/docs/simulation/tutorials/extending-an-epidemic-model/README.md
Comment (nonparibus) [inline resources/docs/simulation/tutorials/extending-an-epidemic-model/README.md:12]:
```suggestion
Start by opening our base template in either [JavaScript](/@hash/getting-started-base) or [Python](/@hash/getting-started-base-python) and forking it into your namespace.
```

hDrive has been deprecated. This shouldn't still be in the docs.

File: resources/docs/simulation/tutorials/phase-1-building-a-simple-hotelling-model-in-2d/query-customers.md
Comment (nonparibus) [inline resources/docs/simulation/tutorials/phase-1-building-a-simple-hotelling-model-in-2d/query-customers.md:21]:
```suggestion
This action can be split into three separate functions within `business.js`: `send_message`, `price_messaging`, `movement_messaging`.
```

**Review by nonparibus — COMMENTED** [review]
Left some minor suggestions to be approved before merging. Am leaving functional UAT/QA to check out rendered pages to another reviewer.

File: resources/docs/simulation/creating-simulations/collaboration/README.md
Comment (akash-joshi) [reply resources/docs/simulation/creating-simulations/collaboration/README.md:30]: This is the rule which is causing it: https://github.com/DavidAnson/markdownlint/blob/main/doc/Rules.md#md026---trailing-punctuation-in-heading

Let me know if we want to disable it.

Also, thanks for going through all the changes! I know there are a lot.

File: .github/workflows/markdown-linting/package.json
Comment (kachkaev) [inline .github/workflows/markdown-linting/package.json:9]: How about putting these at the top of the repo (in this PR or later)? Having  `markdownlint-cli` and `.markdownlint.json` there will let folks use https://marketplace.visualstudio.com/items?itemName=DavidAnson.vscode-markdownlint and thus see the issues before a commit is made.

File: .github/workflows/markdown-linting/.markdownlint.json
Comment (kachkaev) [reply .github/workflows/markdown-linting/.markdownlint.json:2]: Starting with a small number of rules and then extending it is a good strategy, so happy for us to disable rules that would make git diff too massive 👍

File: .github/workflows/mdx-lint.yml
Comment (kachkaev) [reply .github/workflows/mdx-lint.yml:5]: Interesting. I thought that CI would be triggered anyway because of 

```
pull_request:
    branches: [ master ]
```

**Review by kachkaev — COMMENTED** [review]
👍 Nice first step in the right direction.

Do we run Prettier for all our markdowns? If we don’t, we can configure this in a follow-up PR. Having Prettier will help reduce the number `markdownlint` rules we'll need (similar to ESLint + Prettier combo).

File: .github/workflows/markdown-linting/.markdownlint.json
Comment (CiaranMn) [inline .github/workflows/markdown-linting/.markdownlint.json:3]: I'm curious on this one - what are the failures? Is it inside code or something?

File: .github/workflows/mdx-lint.yml
Comment (CiaranMn) [reply .github/workflows/mdx-lint.yml:5]: There is no `master` branch in this repository - this (and below) should be `main`

File: .github/workflows/markdown-linting/package.json
Comment (CiaranMn) [reply .github/workflows/markdown-linting/package.json:9]: Agree with this, and I also think we might as well also format / lint all the markdown in the repository rather than just the resources folder.

Can you also please use `yarn` to install rather than `npm`? So we don't end up with a mixture of lock files in the repo - we'll be bringing in `yarn.lock`s from hash.dev. Can you also create a task to install with `yarn` for the algolia action too in a separate PR? We have a `package-lock.json` over there.

**Review by CiaranMn — CHANGES_REQUESTED** [review]
Thanks @akash-joshi - I agree with Alex that we can have this at the root and lint all markdown, since we have a few readmes too.

File: .github/workflows/markdown-linting/.markdownlint.json
Comment (kachkaev) [reply .github/workflows/markdown-linting/.markdownlint.json:3]: By default, `markdownlint` wants us to replace `hello https://example.com world` with `hello <https://example.com> world`, assuming that URL parsing may be unstable (e.g. when there is `?` at the very end or some UTF8 chars in the middle). Rule details can be found here:
https://github.com/updownpress/markdown-lint/blob/master/rules/034-no-bare-urls.md

I tried living both with and without this rule, but TBH did not find any practical value in it. When the rule is enabled, contributors get confused because wrapping URLs with `<>` is not that common or intuitive.

IMHO we can start with the rule being disabled and reconsider the decision later on if raw links bring issues.

File: .github/workflows/mdx-lint.yml
Comment (kachkaev) [reply .github/workflows/mdx-lint.yml:5]: Good point! I thought we still used `master` for historical reasons. Then we probably want something like:

```yml
on:
  push:
    branches:
      - main
  pull_request:
    branches:
      - main
```

File: resources/docs/simulation/creating-simulations/collaboration/README.md
Comment (kachkaev) [reply resources/docs/simulation/creating-simulations/collaboration/README.md:30]: We can probably disable this rule for now. Having `!`, `...` or `?` at the end of the headers is fine (but we may want to disable `;` `,` etc. in the future when we have more time to thing of the rules)

File: .github/workflows/markdown-linting/package.json
Comment (akash-joshi) [reply .github/workflows/markdown-linting/package.json:9]: This was discussed on a call with Alex yesterday, since we don't want a `package.json` at the root of our repo, it's fine to keep it here for now.

Comment (akash-joshi) [reply .github/workflows/markdown-linting/package.json]: This should be fixed now

File: .github/workflows/markdown-linting/.markdownlint.json
Comment (akash-joshi) [reply .github/workflows/markdown-linting/.markdownlint.json]: Using `<>` also prevents the links from being rendered as links, if I'm not wrong.

File: .github/workflows/mdx-lint.yml
Comment (kachkaev) [inline .github/workflows/mdx-lint.yml:20]: How about using yarn here too? A top-level `yarn.lock` file will save us from accidental breaking changes in newly released dependencies.

File: .markdownlintignore
Comment (kachkaev) [inline .markdownlintignore:1]: Does markdonwlint have default extensions to pick? If list the same files as in `.gitignore` and run `markdownlint .`, we might end up running markdownlint for anything including javsacript files, images etc.

File: .markdownlint.json
Comment (kachkaev) [inline .markdownlint.json:12]: How about adding:

```txt
  "list-marker-space": { "ol_multi": 2, "ol_single": 2 },
  "ol-prefix": { "style": "one" }
```
?

I was thinking of suggesting them later, but because we have changes in a lot of lists anyway, now can be a good opportunity. If adding the rules does not make `git diff` much worse, we can save ourselves from ‘breaking’ `git blame` again in the future.

File: .github/workflows/mdx-lint.yml
Comment (akash-joshi) [reply .github/workflows/mdx-lint.yml:20]: We would need to install `yarn` beforehand for that. I think we're using `npm` elsewhere in our CI too, but I might be wrong.

CC: @CiaranMn

Comment (akash-joshi) [reply .github/workflows/mdx-lint.yml:20]: I guess we can add a follow-up task to bring all dependencies to the parent `package.json` and utilize this package: https://github.com/marketplace/actions/github-action-for-yarn

File: package.json
Comment (kachkaev) [inline package.json:2]: We don’t need to specify attributes like `name`, `version`, `description` etc. if we don’t plan to treat this folder as an npm package (i.e. we don’t need to import from it or publish it to the registry). We can go for `"private": true` instead – that’s enough to silence npm/yarn warnings about missing metadata.

File: resources/docs/simulation/concepts/design-considerations/README.md
Comment (kachkaev) [inline resources/docs/simulation/concepts/design-considerations/README.md:81]: Prettier would normally change this to `-`. Might be worth introducing it too to avoid ‘breaking’ `git blame` again in the future.

Comment (akash-joshi) [reply resources/docs/simulation/concepts/design-considerations/README.md:81]: @kachkaev tbh I'm not sure which part you're referring to here. The commit you're referring to is removing a trailing space from the sentence.

File: .markdownlint.json
Comment (akash-joshi) [reply .markdownlint.json:13]: I tried this, but adding `"ol-prefix": { "style": "one" }` will require manual edits in a lot of files, so I recommend fixing it later as we edit those files.

Comment (akash-joshi) [reply .markdownlint.json:13]: Also I recommend local testing after making this change, since I'm not sure yet how our specific mdx-parser will handle this change so capturing it as a separate task: https://app.asana.com/0/1199550852792212/1201333602499681/f

File: resources/docs/simulation/concepts/design-considerations/README.md
Comment (kachkaev) [reply resources/docs/simulation/concepts/design-considerations/README.md:81]: We are currently using markdownlint without prettier and can make our files even cleaner if we combine these tools. Prettier will replace `*` with `-` when we add it. If we do this shortly after merging this PR, our `git blame` won't be polluted by non-essential changes later.

**Review by kachkaev — APPROVED** [review]
👍👍👍

Here are some possible next steps:

- Add `yarn lint:prettier` and `yarn fix:prettier` (e.g. to replace `*` with `-` in ordered lists)
- Avoid multiple  `yarn.lock` files and ensure we call `yarn` instead of `npm` in CI
- Switch from `1.2.3.` to `1.1.1.` in ordered lists and ensure we offset all ordered lists by four chars instead of three (otherwise, nested ` ``` ` blocks are harder to deal with)

Happy to help!

**Comment by akash-joshi** [comment]
@kachkaev adding the next steps as follow-up tasks:
- Add yarn lint:prettier and yarn fix:prettier (e.g. to replace * with - in ordered lists) - https://app.asana.com/0/1199550852792212/1201333602499697/f
- Avoid multiple yarn.lock files and ensure we call yarn instead of npm in CI - https://app.asana.com/0/1199550852792212/1201333602499699/f

**Review by CiaranMn — APPROVED** [review]

**Pull request merged**
